Source/WebCore/accessibility/AXLiveRegionManager.cpp

```diff
@@ -0,0 +1,317 @@
+            result.append("additions"_s);
+            first = false;
+        }
+        if (snapshot.liveRegionRelevant.contains(LiveRegionRelevant::Removals)) {
+            if (!first)
+                result.append(' ');
+            result.append("removals"_s);
+            first = false;
+        }
+        if (snapshot.liveRegionRelevant.contains(LiveRegionRelevant::Text)) {
+            if (!first)
+                result.append(' ');
+            result.append("text"_s);
+            first = false;
+        }
+        if (snapshot.liveRegionRelevant.contains(LiveRegionRelevant::All)) {
+            if (!first)
+                result.append(' ');
+            result.append("all"_s);
+        }
+    }
+    result.append('\n');
+
+    result.append("\tObjects: "_s);
+    result.append(snapshot.objects.size());
+    result.append('\n');
+
+    for (size_t i = 0; i < snapshot.objects.size(); ++i) {
+        const auto& object = snapshot.objects[i];
+        result.append("\t\t["_s);
+        result.append(i);
+        result.append("] AXID="_s);
+        result.append(object.objectID.loggingString());
+        result.append(" text=\""_s);
+        result.append(object.text);
+        result.append("\"\n"_s);
+    }
+
+    return result.toString();
+}
+
+void AXLiveRegionManager::registerLiveRegion(AccessibilityObject& object, bool speakIfNecessary)
+{
+    auto axID = object.objectID();
+
+    m_liveRegions.set(axID, buildLiveRegionSnapshot(object));
+    // Alerts should speak when added to the page (or initialized for the first time), unlike all other live regions.
+    bool isAlertOrAlertDialog = speakIfNecessary && (object.role() == AccessibilityRole::ApplicationAlert || object.role() == AccessibilityRole::ApplicationAlertDialog);
+    if (isAlertOrAlertDialog)
+        handleLiveRegionChange(object, true);
+}
+
+void AXLiveRegionManager::unregisterLiveRegion(AXID axID)
+{
+    m_liveRegions.remove(axID);
+}
+
+static LiveRegionStatus stringToLiveRegionStatus(const String& string)
+{
+    if (string == "assertive")
+        return LiveRegionStatus::Assertive;
+    if (string == "polite")
+        return LiveRegionStatus::Polite;
+
+    return LiveRegionStatus::Off;
+}
+
+static OptionSet<LiveRegionRelevant> stringToLiveRegionRelevant(const String& string)
+{
+    Vector<String> strings = string.split(" "_s);
+    OptionSet<LiveRegionRelevant> result;
+    for (const auto& attribute : strings) {
+        if (attribute == "additions")
+            result.add(LiveRegionRelevant::Additions);
+        else if (attribute == "all")
+            result.add(LiveRegionRelevant::All);
+        else if (attribute == "removals")
+            result.add(LiveRegionRelevant::Removals);
+        else if (attribute == "text")
+            result.add(LiveRegionRelevant::Text);
+    }
+    return result;
+}
+
+void AXLiveRegionManager::handleLiveRegionChange(AccessibilityObject& object, bool speakAll)
```

Comment:
How about an enum instead of bool here? Values like AnnounceEntireRegion vs AnnounceChangesOnly

Also let's prefer verbs like "announce" because live regions are rendered in braille too, not just speech